### PR TITLE
Use postgres for persistence over google sheets, other minor improvements

### LINKED ISSUE #1: General idea dump
## Submissions
* Priority is to not lose submissions
* People should be able to follow a "convesation" to submit one artwork. What they need to submit:
  * Preview of the work (a still image, blurred if NSFW).
  * Title of the work.
  * Link to the complete work (if possible a Telegram link).
  * List of potential CNs (if NSFW).
  * @ from the account or channel or both.
* People must accept the rules before submitting.
* For now, submissions are a) stored locally in a .csv b) stored in a Google Sheet c) shared on a channel dump.
* Ideally, we gon stop using Google.
* However, submissions should still be dumped in a channel, so that we can share them to the public once the competition is over.
* A lot of the text can/should be sourced _not_ directly in the code.
* Cancelling should be possible at any time.
* Recovery should be possible at any time.
* A sumup should be submitted to the person before validating the submission.

## Vote
* A selected number of people, the jury, should be able to vote.
* For that, they should be able to:
  * Attribute one or multiple categories for each submission.
  * Attribute one vote between 1 and 5 for each category for each submission.
* Only admins should be able to vote in the first place.
* At the end of the vote period, we should have a leaderboard per category.
* It was discussed to have a "public vote" at some point, but reactions on the channel do the trick. It sounds like to much energy to enable people to vote in the bot, it iss not a priority, and it will probably get less usage than channel reactions.

## Communication
* People should be able to communicate with the admins through the bot.
* For now, each message outside a conversation that is not a command is forwarded to the admin group.
* If a reply is made to the forwarded message, that reply is sent to the person.
* It works a bit like @Livegrambot for now.

### PR DESCRIPTION
**Pull request by oohwooh**
closes #1

### CODE REVIEW

**Comment by oohwooh** [comment]
remaining todo: 

- [ ] add ability to create `Contest`s and `Theme`s
- [ ] give the most leeway possible for people submitting (is it before the end date anywhere on earth)
- [ ] voting
- [ ] deploy instructions because stache is a documentation freak and will execute me if i don't properly document my code

File: requirements.txt [hunk 1]
```diff
@@ -1,7 +1,43 @@
...
+six==1.16.0
+toml==0.10.2
+tomli==2.0.1
+tornado==6.2
+typing_extensions==4.4.0
+tzdata==2022.7
+tzlocal==4.2
+uritemplate==4.1.1
+urllib3==1.26.14
+virtualenv==20.17.1
```
Comment (Amustache) [inline requirements.txt:43]: That's a hell of a lot of dependencies

File: main.py [hunk 2]
```diff
@@ -369,16 +410,17 @@ def main() -> None:
...
 
     dispatcher.add_handler(conv_handler)
     dispatcher.add_handler(CommandHandler(["notify", "notify_all"], notify_all))
+    dispatcher.add_handler(CommandHandler(["admin", "add_admin", "allow"], add_admin))
     dispatcher.add_handler(MessageHandler(Filters.chat_type.private & ~Filters.command, forward_to_chat))
     dispatcher.add_handler(
         MessageHandler(Filters.chat(ADMINS_GROUPCHAT) & Filters.reply & ~Filters.command, forward_to_user)
     )
-
+    dispatcher.add_error_handler(error_handler)
```
Comment (Amustache) [inline main.py:455]: O:)

File: main.py [hunk 3]
```diff
@@ -34,39 +32,25 @@
...
-service = build("sheets", "v4", credentials=creds)
-
 
 def start(update: Update, context: CallbackContext) -> int:
-    user = update.effective_user
-    with open(userids_file, "a") as f:
-        f.write("{}\n".format(user.id))
-
+    db.connect(reuse_if_open=True)
+    user = User.get_or_create(telegramId=str(update.effective_user.id))[0]
```
Comment (Amustache) [inline main.py]:
```suggestion
    user, _ = User.get_or_create(telegramId=str(update.effective_user.id))

```

File: main.py [hunk 4]
```diff
@@ -75,6 +59,13 @@ def start(update: Update, context: CallbackContext) -> int:
...
+    db.connect(reuse_if_open=True)
+    current_contest = Contest.get_or_none(Contest.starts <= datetime.now(), Contest.ends >= datetime.now())
+    db.close()
+    if current_contest is None:
+        update.message.reply_text("I'm sorry, however submissions are not currently open for TADA.\n"
+                                  "If you feel this is in error, please send a message to this bot to get in contact with the admins.")
+        return ConversationHandler.END
     update.message.reply_text("*Rules*", parse_mode="MarkdownV2")
     update.message.reply_text(
         "- Submissions will be open from February, 1st to February, 28th.\n"
```
Comment (Amustache) [inline main.py:72]: Maybe use `Contest.starts`/`Contest.ends`?

File: main.py [hunk 5]
```diff
@@ -124,8 +115,15 @@ def submit_photo(update: Update, context: CallbackContext) -> int:
...
+    db.connect(reuse_if_open=True)
+    current_contest = Contest.get_or_none(Contest.starts <= datetime.now(), Contest.ends >= datetime.now())
+    db.close()
+    if current_contest is None:
+        update.message.reply_text("I'm sorry, however submissions are not currently open for TADA.\n"
+                                  "If you feel this is in error, please send a message to this bot to get in contact with the admins.")
+        return ConversationHandler.END
     update.message.reply_text("Alright, let's start again, then!\n" "Please give use the title of your artwork.\n")
-
+    context.user_data[update.effective_user.id]["nsfw"] = None
```
Comment (Amustache) [inline main.py:127]: lol

File: main.py [hunk 6]
```diff
@@ -233,36 +233,47 @@ def confirmation(update: Update, context: CallbackContext) -> int:
 
 
 def submission(update: Update, context: CallbackContext) -> int:
+    user_data = context.user_data[update.effective_user.id]
+    db.connect(reuse_if_open=True)
+    user = User.get_or_create(telegramId=str(update.effective_user.id))[0]
```
Comment (Amustache) [inline main.py]:
```suggestion
    user, _ = User.get_or_create(telegramId=str(update.effective_user.id))

```

File: main.py [hunk 7]
```diff
@@ -280,50 +291,74 @@ def cancel(update: Update, context: CallbackContext) -> int:
...
-                update.message.reply_text(
-                    "Sent to {} people out of {} ({} failed).".format(total - blocked, total, blocked)
-                )
-            else:
-                text_to_send = "🗣 Message from admin 🗣\n{}".format(update.effective_message.text.split(" ", 1)[1])
-                update.message.reply_text("This is a preview:").reply_text(text_to_send).reply_text(
-                    "Reply /notify to the previous message to send it."
-                )
+    db.connect(reuse_if_open=True)
+    user = User.get_or_create(telegramId=str(update.effective_user.id))[0]
```
Comment (Amustache) [inline main.py]:
```suggestion
    user, _ = User.get_or_create(telegramId=str(update.effective_user.id))

```

File: main.py [hunk 8]
```diff
@@ -332,19 +367,25 @@ def vote(update: Update, context: CallbackContext) -> None:
...
 
-def main() -> None:
-    # Create files if not existing
-    open(userids_file, "a").close()
-    with open(submissions_file, "a", newline="") as csvfile:
-        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
-        writer.writeheader()
+def add_admin(update: Update, context: CallbackContext) -> None:
+    db.connect(reuse_if_open=True)
+    user = User.get_or_create(telegramId=str(update.effective_user.id))[0]
```
Comment (Amustache) [inline main.py]:
```suggestion
    user, _ = User.get_or_create(telegramId=str(update.effective_user.id))

```

File: main.py [hunk 9]
```diff
@@ -332,19 +367,25 @@ def vote(update: Update, context: CallbackContext) -> None:
...
+    if user.isAdmin:
+        if update.message.reply_to_message:
+            newAdmin = User.get_or_create(telegramId=update.message.reply_to_message.from_user.id)[0]
```
Comment (Amustache) [inline main.py]:
```suggestion
            newAdmin, _ = User.get_or_create(telegramId=update.message.reply_to_message.from_user.id)

```

File: db/models.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    telegramId = TextField(unique=True)
+    isAdmin = BooleanField(default=False)
+
+
+class Submission(BaseModel):
+    filename = TextField()
+    title = TextField()
+    link = TextField()
+    nsfw = BooleanField()
+    contentWarnings = TextField()
```
Comment (Amustache) [inline db/models.py:29]: Maybe only use contentWarnings as a bool? As if, `null=True`, and if it's empty, it's not nsfw

File: db/models.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    at = TextField()
+    author = TextField()
+    contest = ForeignKeyField(Contest, backref='submissions')
+    user = ForeignKeyField(User, backref='submissions')
```
Comment (Amustache) [inline db/models.py:33]: I like that

**Review by Amustache — APPROVED** [review]

**Pull request merged**